mutation_partition.cc

```diff
@@ -305,14 +305,104 @@ stop_iteration mutation_partition::apply_monotonically(const schema& s, mutation
     _static_row.apply_monotonically(s, column_kind::static_column, std::move(p._static_row));
     _static_row_continuous |= p._static_row_continuous;
 
+    rows_entry::tri_compare cmp(s);
+    auto del = current_deleter<rows_entry>();
+
+    // Compacts rows in [i, end) with the tombstone.
+    // Erases entries which are left empty by compaction.
+    // Does not affect continuity.
+    auto apply_tombstone_to_rows = [&] (apply_resume::stage stage, tombstone tomb, rows_type::iterator i, rows_type::iterator end) -> stop_iteration {
+        if (!preemptible) {
+            // Compaction is attempted only in preemptible contexts because it can be expensive to perform and is not
+            // necessary for correctness.
+            return stop_iteration::yes;
+        }
+
+        while (i != end) {
+            rows_entry& e = *i;
+            can_gc_fn never_gc = [](tombstone) { return false; };
+
+            bool all_dead = e.dummy() || !e.row().compact_and_expire(s,
+                                                                     tomb,
+                                                                     gc_clock::time_point::min(),  // no TTL expiration
+                                                                     never_gc,                     // no GC
+                                                                     gc_clock::time_point::min()); // no GC
+
+            auto next_i = std::next(i);
+            bool inside_continuous_range = !tracker ||
+                    (e.continuous() && (next_i != _rows.end() && next_i->continuous()));
+
+            if (all_dead && e.row().empty() && inside_continuous_range) {
+                i = _rows.erase(i);
+                if (tracker) {
+                    tracker->on_remove();
+                }
+                del(&e);
+            } else {
+                i = next_i;
+            }
+
+            if (need_preempt() && i != end) {
+                res = apply_resume(stage, position_in_partition(i->position()));
```

Comment:
What happens if copying causes us to throw here?

I'm guessing we'll be left in some undefined state. Maybe better to swallow the exception and not stop in that case.